include/zephyr/spinlock.h

```diff
@@ -78,7 +78,7 @@ struct k_spinlock {
 #endif /* CONFIG_SPIN_LOCK_TIME_LIMIT */
 #endif /* CONFIG_SPIN_VALIDATE */
 
-#if defined(CONFIG_CPP) && !defined(CONFIG_SMP) && \
+#if (defined(CONFIG_CPP) || defined(CONFIG_RUST)) && !defined(CONFIG_SMP) && \
```

Comment:
FWIW we really need to come up with a clean solution here to the "zero length spinlock" problem.  The waste of the byte (plus alignment, so generally a word) is bad enough, but the fact that the ABI changed depending on whether you want C++ (or, now, Rust) support is a pretty terrible wart.  Spinlock records in outer struts really do need to vanish[1] and not be a degenerate 1-byte struct.

That said, it's not clear to me why Rust would need this?  Does rust disallow empty struts too?

[1] On single-cpu systems without SPIN_VALIDATE, that is; i.e. the circumstances where a spinlock can be implemented with a bare arch_irq_lock/unlock().

Reply:
Rust is just fine with empty structs, in C its an extension and likely makes things unhappy when generating bindings

Reply:
Rust is "fine" with it, but every struct containing one of them will generate a warning in every API call using it about it not being FFI-Safe. I'm not sure what the right solution is here.

Reply:
> FWIW we really need to come up with a clean solution here to the "zero length spinlock" problem. 

@andyross Is it possible to use a singleton for that purpose?

I actually had a bug open for that for a long time and IIRC just closed it thinking it was long forgotten / -ENOTIME.

Reply:
Does a zero-length array at the end of a struct avoid having to pad the struct with 1 byte?

Reply:
I see a couple possible solutions here:
1. Leave the dummy byte in the spinlock.
2. Leave the empty spinlock, and suppress the warning in the generated rust code, but at a whole level. This will remove the warning, but possibly miss any other warnings about FFI issues.
3. Enhance bindgen to be able to insert the warning suppression just on specific types.

Thoughts?

Reply:
For now, I've gone with solution 2. I've reverted the change to the C code (and will remove the patches entirely before final review), and suppressed the warning. One suggestion I did see was to add code to edit the generated bindings before using them. This is definitely possible, but a bit more awkward and fragile, and seems riskier to me than just suppressing the warning for all of the binding code. The bindings are generated from C structs so should always be valid as arguments to C functions.

Reply:
That sounds best to me from a general compatibility standpoint, though the C++ wart still exists and needs some genius to come along and figure out how to make this work compatibly with a clean API.  But that bikeshed isn't relevant to this PR.  +1 from me, though obviously I lack the Rust-fu to do much review on the broader PR.